In [52]:
!pip install ucimlrepo

In [111]:
# Importa a biblioteca pandas para manipulação de dados em formato tabular
import pandas as pd

# Importa a função fetch_ucirepo da biblioteca ucimlrepo para buscar datasets do UCI Machine Learning Repository
from ucimlrepo import fetch_ucirepo

# Faz o download do conjunto de dados 'Breast Cancer Wisconsin (Diagnostic)' usando o ID 17 do repositório
breast_cancer_wisconsin_diagnostic = fetch_ucirepo(id=17)

# Extrai os dados das características (features) do conjunto de dados
X = breast_cancer_wisconsin_diagnostic.data.features

# Extrai os dados dos rótulos (targets), ou seja, o diagnóstico de câncer
Y = breast_cancer_wisconsin_diagnostic.data.targets

In [112]:
display(Y)

Diagnosis
0           M
1           M
2           M
3           M
4           M
..        ...
564         M
565         M
566         M
567         M
568         B

[569 rows x 1 columns]

In [113]:
display(X)


radius1  texture1  perimeter1   area1  smoothness1  compactness1  \
0      17.99     10.38      122.80  1001.0      0.11840       0.27760   
1      20.57     17.77      132.90  1326.0      0.08474       0.07864   
2      19.69     21.25      130.00  1203.0      0.10960       0.15990   
3      11.42     20.38       77.58   386.1      0.14250       0.28390   
4      20.29     14.34      135.10  1297.0      0.10030       0.13280   
..       ...       ...         ...     ...          ...           ...   
564    21.56     22.39      142.00  1479.0      0.11100       0.11590   
565    20.13     28.25      131.20  1261.0      0.09780       0.10340   
566    16.60     28.08      108.30   858.1      0.08455       0.10230   
567    20.60     29.33      140.10  1265.0      0.11780       0.27700   
568     7.76     24.54       47.92   181.0      0.05263       0.04362   

     concavity1  concave_points1  symmetry1  fractal_dimension1  ...  radius3  \
0       0.30010          0.14710     0.2419             0.07871  ...   25.380   
1       0.08690          0.07017     0.1812             0.05667  ...   24.990   
2       0.19740          0.12790     0.2069             0.05999  ...   23.570   
3       0.24140          0.10520     0.2597             0.09744  ...   14.910   
4       0.19800          0.10430     0.1809             0.05883  ...   22.540   
..          ...              ...        ...                 ...  ...      ...   
564     0.24390          0.13890     0.1726             0.05623  ...   25.450   
565     0.14400          0.09791     0.1752             0.05533  ...   23.690   
566     0.09251          0.05302     0.1590             0.05648  ...   18.980   
567     0.35140          0.15200     0.2397             0.07016  ...   25.740   
568     0.00000          0.00000     0.1587             0.05884  ...    9.456   

     texture3  perimeter3   area3  smoothness3  compactness3  concavity3  \
0       17.33      184.60  2019.0      0.16220       0.66560      0.7119   
1       23.41      158.80  1956.0      0.12380       0.18660      0.2416   
2       25.53      152.50  1709.0      0.14440       0.42450      0.4504   
3       26.50       98.87   567.7      0.20980       0.86630      0.6869   
4       16.67      152.20  1575.0      0.13740       0.20500      0.4000   
..        ...         ...     ...          ...           ...         ...   
564     26.40      166.10  2027.0      0.14100       0.21130      0.4107   
565     38.25      155.00  1731.0      0.11660       0.19220      0.3215   
566     34.12      126.70  1124.0      0.11390       0.30940      0.3403   
567     39.42      184.60  1821.0      0.16500       0.86810      0.9387   
568     30.37       59.16   268.6      0.08996       0.06444      0.0000   

     concave_points3  symmetry3  fractal_dimension3  
0             0.2654     0.4601             0.11890  
1             0.1860     0.2750             0.08902  
2             0.2430     0.3613             0.08758  
3             0.2575     0.6638             0.17300  
4             0.1625     0.2364             0.07678  
..               ...        ...                 ...  
564           0.2216     0.2060             0.07115  
565           0.1628     0.2572             0.06637  
566           0.1418     0.2218             0.07820  
567           0.2650     0.4087             0.12400  
568           0.0000     0.2871             0.07039  

[569 rows x 30 columns]

In [114]:
import sklearn
from sklearn.model_selection import train_test_split

In [115]:
X_treinamento, X_teste, Y_treinamento, Y_teste = train_test_split(X, Y, test_size=0.25)

In [117]:
# Substitui os rótulos da coluna 'Diagnosis' no conjunto de teste.
# Converte 'M' (Maligno) para 1 e 'B' (Benigno) para 0, facilitando o uso em algoritmos de machine learning.
Y_teste = Y_teste.replace({'Diagnosis': {'M': 1, 'B': 0}})

# Faz a mesma substituição no conjunto de treinamento: 'M' vira 1, 'B' vira 0.
Y_treinamento = Y_treinamento.replace({'Diagnosis': {'M': 1, 'B': 0}})

# Imprime as formas (shape) das matrizes de entrada e saída do treinamento — útil para verificar se os dados estão coerentes.
print(f'{X_treinamento.shape, Y_treinamento.shape}\n')

# Imprime as formas dos dados de teste — ajuda a validar a divisão dos dados e visualizar o tamanho.
print(f'{X_teste.shape, Y_teste.shape}')

((426, 30), (426, 1))

((143, 30), (143, 1))


In [118]:
X_treinamento.shape, Y_treinamento.shape

((426, 30), (426, 1))

In [119]:
X_teste.shape, Y_teste.shape

((143, 30), (143, 1))

In [120]:
import tensorflow as tf
from tensorflow.keras.models import Sequential

Após importação da biblioteca tensorflow é possível criar a rede neural, aqui é criado uma camada de entrada com formato de 30, pois são 30 caracteristicas que serão analisadas, a segunda camada densa com 16 unidades e a ultima camada a de saída com a função sigmoide.

In [99]:
# Cria uma rede neural sequencial — ou seja, onde cada camada é adicionada uma após a outra.
rede_neural = Sequential([

    # Define a camada de entrada com 30 características (features) — isso corresponde ao número de colunas nos dados.
    tf.keras.layers.InputLayer(shape = (30,)),

    # Adiciona uma camada densa (totalmente conectada) com 16 neurônios.
    # A função de ativação 'relu' ajuda a introduzir não-linearidade no modelo.
    # O inicializador 'random_uniform' define os pesos iniciais aleatoriamente, em uma distribuição uniforme.
    tf.keras.layers.Dense(units = 16, activation='relu', kernel_initializer='random_uniform'),


    # Adiciona a camada de saída com 1 neurônio, ideal para problemas binários (diagnóstico de câncer: maligno ou benigno).
    # A função de ativação 'sigmoid' retorna um valor entre 0 e 1, representando a probabilidade.
    tf.keras.layers.Dense(units = 1, activation='sigmoid')
])

In [121]:
rede_neural.summary()

Model: "sequential_7"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_18 (Dense)                │ (None, 16)             │           496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_19 (Dense)                │ (None, 16)             │           272 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_20 (Dense)                │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,357 (9.21 KB)

 Trainable params: 785 (3.07 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 1,572 (6.14 KB)

1. Model: "sequential"
Indica que o modelo foi construído como uma sequência linear de camadas, ou seja, cada camada recebe os dados da camada anterior.

2. Layer (type)
Aqui são listadas as camadas do modelo. No seu caso, duas delas são Dense, ou seja, totalmente conectadas:

A primeira é a camada oculta com 16 neurônios.

A segunda é a camada de saída, com apenas 1 neurônio para classificar o resultado.

3. Output Shape
Mostra a forma da saída de cada camada:

(None, 16) significa que a camada oculta produzirá 16 valores por entrada. O None representa o número de amostras (exemplos), que pode variar.

(None, 1) indica que a camada de saída produzirá 1 valor por entrada — ideal para classificação binária (ex: câncer benigno ou maligno).

4. Param #
Exibe o número de parâmetros treináveis da camada:

A primeira camada possui 496 parâmetros, que são os pesos (30 inputs × 16 neurônios = 480) e os 16 bias.

A segunda camada tem 17 parâmetros: 16 pesos (de cada neurônio anterior) + 1 bias.

5. Total params
É a soma de todos os parâmetros treináveis do modelo, que dá 513.

6. Trainable params
Quantidade de parâmetros que serão atualizados durante o treinamento (também 513, pois todos são treináveis).

7. Non-trainable params
Esses seriam os parâmetros que o modelo não irá ajustar, mas como você não usou nenhuma camada com congelamento (como camadas pré-treinadas), o valor é 0.

In [122]:
# Compila o modelo definindo como ele será treinado.
# Isso inclui o algoritmo de otimização, a função de perda e as métricas de avaliação.

rede_neural.compile(

    # Otimizador 'adam' é um dos mais usados — ele ajusta os pesos automaticamente de forma eficiente.
    optimizer="adam",

    # A função de perda 'binary_crossentropy' mede o erro da rede em problemas de classificação binária.
    # Ela compara a saída prevista com o rótulo verdadeiro (0 ou 1), ajustando os pesos para minimizar esse erro.
    loss="binary_crossentropy",

    # Métrica 'binary_accuracy' calcula a proporção de previsões corretas (acertos) no treino e teste.
    # Assim você consegue acompanhar o desempenho da rede ao longo do treinamento.
    metrics=["binary_accuracy"]
)

In [123]:
rede_neural.fit(X_treinamento, Y_treinamento, batch_size = 10, epochs=100)

Epoch 1/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - binary_accuracy: 0.4754 - loss: 0.7069
Epoch 2/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - binary_accuracy: 0.8116 - loss: 0.5331 
Epoch 3/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.9021 - loss: 0.3458
Epoch 4/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8833 - loss: 0.3074
Epoch 5/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.9094 - loss: 0.2500
Epoch 6/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8627 - loss: 0.3339
Epoch 7/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8887 - loss: 0.2614
Epoch 8/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.9004 - loss: 0.2281
Epoch 9/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8720 - loss: 0.2606
Epoch 10/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8881 - loss: 0.2711
Epoch 11/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.9037 

Teste para ver se a rede neural acerta os resultados com os dados de testes, e a transformação dos valores em verdadeiro ou falso

In [125]:
previsoes = rede_neural.predict(X_teste)
previsoes = (previsoes > 0.5)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 


In [126]:
previsoes

array([[False],
       [ True],
       [ True],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [

In [127]:
Y_teste

Diagnosis
511          0
161          1
85           1
484          0
459          0
..         ...
193          1
528          0
461          1
361          0
539          0

[143 rows x 1 columns]

In [128]:
from sklearn.metrics import accuracy_score, confusion_matrix


Após importar a biblioteca podemos ver que a porcentagem de acerto da rede neural é de 93%

In [129]:
accuracy_score(Y_teste, previsoes)

0.9300699300699301

Matriz de confusão nos mostra que 84 benignos foram classificados como benignos e 4 que era benigno foi classificado como maligno, e nos mostra que 6 malignos foram classificados como benignos e 49 malignos foram classificados como malignos. A diagonal principal nos mostra a quantidade de acertos.

In [131]:
confusion_matrix(Y_teste, previsoes)

array([[93,  0],
       [10, 40]])

In [132]:
84 + 49, 4 + 6

(133, 10)

In [133]:
rede_neural.evaluate(X_teste, Y_teste)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - binary_accuracy: 0.9333 - loss: 0.1406  


[0.14481264352798462, 0.9300699234008789]

Criação de mais uma camada oculta.

In [134]:
rede_neural = Sequential([
    tf.keras.layers.InputLayer(shape = (30,)),
    tf.keras.layers.Dense(units = 16, activation = 'relu', kernel_initializer = 'random_uniform'),
    tf.keras.layers.Dense(units = 16, activation = 'relu', kernel_initializer = 'random_uniform'),
    tf.keras.layers.Dense(units = 1, activation = 'sigmoid')
])

In [135]:
rede_neural.summary()

Model: "sequential_8"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_21 (Dense)                │ (None, 16)             │           496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_22 (Dense)                │ (None, 16)             │           272 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_23 (Dense)                │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 785 (3.07 KB)

 Trainable params: 785 (3.07 KB)

 Non-trainable params: 0 (0.00 B)

Criação de um otimizador que limita os pesos a 0.5

In [136]:
otimizador = tf.keras.optimizers.Adam(learning_rate = 0.001, clipvalue= 0.5)

In [137]:
rede_neural.compile(optimizer = otimizador, loss = 'binary_crossentropy', metrics = ['binary_accuracy'])

In [138]:
rede_neural.fit(X_treinamento, Y_treinamento, batch_size = 10, epochs=100)

Epoch 1/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - binary_accuracy: 0.5083 - loss: 0.7413
Epoch 2/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - binary_accuracy: 0.6941 - loss: 0.5890
Epoch 3/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - binary_accuracy: 0.8484 - loss: 0.4717
Epoch 4/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - binary_accuracy: 0.8189 - loss: 0.4249
Epoch 5/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - binary_accuracy: 0.8364 - loss: 0.3619
Epoch 6/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - binary_accuracy: 0.8982 - loss: 0.2992
Epoch 7/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - binary_accuracy: 0.8767 - loss: 0.2823
Epoch 8/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.9075 - loss: 0.2537
Epoch 9/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.9138 - loss: 0.2299
Epoch 10/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8973 - loss: 0.2612
Epoch 11/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.9217 

Mostra o valor de erro e da acuracia.

In [140]:
rede_neural.evaluate(X_teste, Y_teste)

5/5 ━━━━━━━━━━━━━━━━━━━━ 1s 26ms/step - binary_accuracy: 0.9059 - loss: 0.2271 


[0.22440789639949799, 0.9090909361839294]

Acessando os pesos da primeira camada da rede neural.

In [141]:
pesos0 = rede_neural.layers[0].get_weights()

São 2 arrays pois mostra a conexao dos bias e a conexao de cada neuronio.

In [142]:
len(pesos0)

2

Bias

In [143]:
len(pesos0[1])

16

Camada

In [144]:
len(pesos0[0])

30

In [145]:
pesos1 = rede_neural.layers[1].get_weights()
pesos1

[array([[ 7.32779503e-03,  2.51598023e-02, -4.74416167e-02,
         -5.94427809e-03, -2.44779866e-02, -7.46315205e-03,
          9.40982029e-02, -8.34434107e-03, -1.78458720e-01,
         -1.28202409e-01,  4.44149040e-02,  2.23653652e-02,
         -1.16575629e-01, -4.19271290e-02, -2.69146692e-02,
         -1.45790964e-01],
        [ 3.19474675e-02, -4.60997708e-02, -1.10468492e-02,
          1.17854029e-03,  1.65749677e-02,  1.18358508e-02,
          7.92487711e-03,  1.52011961e-03,  4.95091081e-05,
         -2.78548151e-03,  4.07626070e-02,  4.68724258e-02,
          4.07178290e-02, -3.98190729e-02,  3.93189825e-02,
         -1.29983798e-02],
        [-4.83197458e-02,  2.14874037e-02,  3.54768075e-02,
         -6.87409192e-04,  6.78303093e-03,  2.45146267e-02,
         -9.62923840e-03, -2.87918095e-02, -2.33402494e-02,
         -3.16238776e-02,  2.58462466e-02, -2.06569191e-02,
         -3.55981961e-02,  1.42058022e-02, -1.10955127e-02,
         -4.35527414e-03],
        [ 4.8898372

In [146]:
pesos2 = rede_neural.layers[2].get_weights()
pesos2

[array([[ 0.3428888 ],
        [-0.18503077],
        [ 0.04680698],
        [-0.2537931 ],
        [-0.51957846],
        [ 0.19347402],
        [ 0.93487364],
        [ 0.58801043],
        [-0.04184119],
        [-0.03214225],
        [ 0.404696  ],
        [ 0.25984997],
        [-0.53378814],
        [ 0.3623123 ],
        [ 0.3594558 ],
        [-0.05201236]], dtype=float32),
 array([-0.23868278], dtype=float32)]